scripts/dev/tidy.php

```diff
@@ -0,0 +1,1147 @@
+ */
+function relative(string $file): string
+{
+    $opt = options();
+
+    return str_replace($opt['path'].'/', '', $file);
+}
+
+/**
+ * Overwrite file and save a backup copy.
+ */
+function save(string $file, string $content): bool
+{
+    $opt = options();
+
+    // Backup file
+    if ($opt['backup'] && file_put_contents($file.'~', $content) !== false) {
+        output('`COPY`  '.relative($file).'~: *backup copy saved*');
+    }
+
+    if (file_put_contents($file, $content) === false) {
+        output('**FAIL**  '.relative($file).': **file could not be saved**');
+
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Explode PHP test file sections to array.
+ */
+function getTestSections(string $file): array
+{
+    $fp = fopen($file, 'rb') or error("Cannot open test file: $file");
+
+    $sections = [];
+    $section = '';
+
+    while (!feof($fp)) {
+        $line = fgets($fp);
+
+        if ($line === false) {
+            break;
+        }
+
+        if (preg_match('/^--[_A-Z]+--.*/', $line, $matches)) {
+            $section = $line;
+            $sections[$section] = '';
+        } elseif ($section !== '') {
+            $sections[$section] .= $line;
+        }
+    }
+
+    return $sections;
+}
+
+/**
+ * Merge given PHP test sections back to file contents.
+ */
+function mergeTestSections(array $sections): string
+{
+    $content = '';
+
+    foreach ($sections as $name => $section) {
+        $content .= $name.$section;
+    }
+
+    return $content;
+}
+
+/**
+ * PHP test files are a special case and need special tidying approach for each
+ * section.
+ */
+function tidyPhpTestFile(string $file): string
+{
+    $opt = options();
+    $logs = [];
+    $buffer = [];
+    $sections = getTestSections($file);
+
+    if (!preg_match('/^\-\-TEST\-\-/', key($sections))) {
+        output('**FAIL**  '.relative($file).': **not a PHP test**');
+
+        return false;
+    }
+
+    // Trim trailing whitespace
+    if ($opt['trim_trailing_whitespace']) {
+        foreach ($sections as $name => $section) {
+            $cleanedSection = $section;
+
+            // Trim trailing whitespace after section names
+            $cleanedName = trimTrailingWhitespace($name);
+
+            // Trim trailing whitespace for known sections
+            if (in_array(trim($cleanedName), [
+                '--TEST--',
+                '--DESCRIPTION--',
+                '--SKIPIF--',
+                '--INI--',
+                '--ENV--',
+                '--EXTENSIONS--',
+                '--CREDITS--',
+                '--FILE_EXTERNAL--',
+                ])
+            ) {
+                $cleanedSection = trimTrailingWhitespace($cleanedSection);
+            }
+
+            // Trim trailing whitespace in certain PHP code lines
+            if (in_array(trim($cleanedName), ['--FILE--'])) {
+                $cleanedSection = preg_replace(
+                    '/(*BSR_ANYCRLF)(\<\?php|\?\>|\)[ ]?\{|\)\;)[\t ]+(\R|$)/m',
```

Comment:
Rather than restricting what we trim here, might it make more sense to trim everything but keep a blacklist (or maybe better: Add a `--NO_TIDY--` section to the phpt) of tests that are whitespace sensitive? I think there should be very few tests (things like tests for flexible heredoc/nowdoc) that really need whitespace sensitivity.

Reply:
@nikic rather than adding a specific --NO_TIDY-- section, I would rather go by having an explicit blacklist in the script itself for such exceptions. The tool is written for the tests, not the other way around imo

Reply:
@KalleZ Being formatting-sensitive is a property of the test though ;) Also allows exts to use it.

Reply:
Thats true, I just feel like we should have an actual better way of defining options in tests but thats beyond the scope of this PR. Something like an --OPTIONS-- section or similar that have a similar syntax to --INI-- with test specific options instead of more sections would be a better fit imo.

Reply:
Refactored... The new committed solution now tidies PHP code in tests using two approaches. Depending if the tokenizer extension is available or not:
- regex with some common fixing rules
- or using the tokenizer extension and tidies these normally and more thoroughly. Leaves the Heredoc and multiline strings intact.

I think, important thing is also to have a simple way to write a test. Or to change it in the future. Now, there are two issues in some phpt files, where normal text editors/human typing can't be used for them:
- CR characters are present together with LF. Some editors replace CR with CRLF or LF.... and on top of that they also can't be noticed in text editor if aren't careful. Vim or maybe some hex text editor can do that better though.
- and trailing whitespace (can be avoided with manual editor adjustment or adjusting the .editorconfig)

(depending who you ask of course)